src/runtime_src/core/edge/drm/zocl/zocl_kds.c

```diff
@@ -95,12 +109,135 @@ zocl_ctx_to_info(struct drm_zocl_ctx *args, struct kds_ctx_info *info)
+ *
+ */
+static void
+zocl_remove_client_context(struct drm_zocl_dev *zdev,
+			struct kds_client *client, struct kds_client_ctx *cctx)
+{
+	struct drm_zocl_domain *domain;
+	uuid_t *id = (uuid_t *)cctx->xclbin_id;
+
+	/* Check whether active count exists for this context */
+	if (cctx && !cctx->num_ctx) {
+		/* Get the corresponding domain for this xclbin */
+		domain = zocl_get_domain(zdev, id);
+		if (!domain)
+			return;
+
+		/* Unlock this domain specific xclbin */
+		zocl_unlock_bitstream(domain, id);
+
+		list_del(&cctx->link);
+		if (cctx->xclbin_id)
+			vfree(cctx->xclbin_id);
+		if (cctx)
+			vfree(cctx);
+	}
+}
+
+/*
+ * Create a new client context and lock the bitstrean for the domain
+ * associated with this context.
+ *
+ * @param	zdev:   zocl device structure
+ * @param       client:	KDS client structure
+ * @param       id:	XCLBIN id
+ *
+ * @return      newly created context on success, Error code on failure.
+ *
+ */
+static struct kds_client_ctx *
+zocl_create_client_context(struct drm_zocl_dev *zdev,
+			struct kds_client *client, uuid_t *id)
+{
+	struct drm_zocl_domain *domain;
+	struct kds_client_ctx *cctx;
+	int ret;
+
+	/* Get the corresponding domain for this xclbin */
+	domain = zocl_get_domain(zdev, id);
+	if (!domain)
+		return NULL;
+
+	/* Lock this domain specific xclbin */
+	ret = zocl_lock_bitstream(domain, id);
+	if (ret)
+		return NULL;
+
+	/* Allocate the new client context and store the xclbin */
+	cctx = vzalloc(sizeof(struct kds_client_ctx));
+	if (!cctx) {
+		(void) zocl_unlock_bitstream(domain, id);
+		return NULL;
+	}
+
+	cctx->xclbin_id = vzalloc(sizeof(uuid_t));
+	if (!cctx->xclbin_id) {
+		vfree(cctx);
+		(void) zocl_unlock_bitstream(domain, id);
+		return NULL;
+	}
+	uuid_copy(cctx->xclbin_id, id);
+
+	list_add_tail(&cctx->link, &client->ctx_list);
+
+	return cctx;
+}
+
+/*
+ * Check whether there is a active context for this xclbin in this kds client.
+ *
+ * @param       client:	KDS client structure
+ * @param       id:	XCLBIN id
+ *
+ * @return      existing context on success, NULL on failure.
+ *
+ */
+static struct kds_client_ctx *
+zocl_check_exists_context(struct kds_client *client, uuid_t *id)
+{
+	struct kds_client_ctx *curr;
+
+	/* Find whether the xclbin is already loaded and the context is exists
+	 */
+	list_for_each_entry(curr, &client->ctx_list, link)
+		if (uuid_equal(curr->xclbin_id, id))
+			break;
+
+	/* Not found any matching context */
+	if (&curr->link == &client->ctx_list)
+		return NULL;
+
+	return curr;
+}
+
+/*
+ * Create a new context if no active context present for this xclbin and add
+ * it to the kds.
+ *
+ * @param	zdev:   zocl device structure
+ * @param       client:	KDS client structure
+ * @param       id:	XCLBIN id
+ *
+ * @return      0 on success, error core on failure.
+ *
+ */
 static int
 zocl_add_context(struct drm_zocl_dev *zdev, struct kds_client *client,
 		 struct drm_zocl_ctx *args)
 {
 	struct kds_ctx_info info;
```

Comment:
Uninitialized structure
```suggestion
	struct kds_ctx_info info = {0};
```